sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java

```diff
@@ -0,0 +1,181 @@
+
+  private boolean performanceEnabled = false;
+
+  // WeakHashMap isn't thread safe but ActivityLifecycleCallbacks is only called from the
+  // main-thread
+  private final @NotNull WeakHashMap<Activity, ITransaction> activities = new WeakHashMap<>();
+
+  public ActivityLifecycleIntegration(final @NotNull Application application) {
+    this.application = Objects.requireNonNull(application, "Application is required");
+  }
+
+  @Override
+  public void register(final @NotNull IHub hub, final @NotNull SentryOptions options) {
+    this.options =
+        Objects.requireNonNull(
+            (options instanceof SentryAndroidOptions) ? (SentryAndroidOptions) options : null,
+            "SentryAndroidOptions is required");
+
+    this.hub = Objects.requireNonNull(hub, "Hub is required");
+
+    this.options
+        .getLogger()
+        .log(
+            SentryLevel.DEBUG,
+            "ActivityBreadcrumbsIntegration enabled: %s",
+            this.options.isEnableActivityLifecycleBreadcrumbs());
+
+    performanceEnabled = isPerformanceEnabled(this.options);
+
+    if (this.options.isEnableActivityLifecycleBreadcrumbs() || performanceEnabled) {
+      application.registerActivityLifecycleCallbacks(this);
+      options.getLogger().log(SentryLevel.DEBUG, "ActivityBreadcrumbsIntegration installed.");
+    }
+  }
+
+  private boolean isPerformanceEnabled(final @NotNull SentryAndroidOptions options) {
+    return ((options.getTracesSampleRate() != null || options.getTracesSampler() != null)
+        && options.isEnableAutoActivityLifecycleTracing());
+  }
+
+  @Override
+  public void close() throws IOException {
+    application.unregisterActivityLifecycleCallbacks(this);
+
+    options.getLogger().log(SentryLevel.DEBUG, "ActivityBreadcrumbsIntegration removed.");
+  }
+
+  private void addBreadcrumb(final @NonNull Activity activity, final @NotNull String state) {
+    if (options.isEnableActivityLifecycleBreadcrumbs()) {
+      final Breadcrumb breadcrumb = new Breadcrumb();
+      breadcrumb.setType("navigation");
+      breadcrumb.setData("state", state);
+      breadcrumb.setData("screen", getActivityName(activity));
+      breadcrumb.setCategory("ui.lifecycle");
+      breadcrumb.setLevel(SentryLevel.INFO);
+      hub.addBreadcrumb(breadcrumb);
+    }
+  }
+
+  private @NotNull String getActivityName(final @NonNull Activity activity) {
+    return activity.getClass().getSimpleName();
+  }
+
+  private void startTracing(final @NonNull Activity activity) {
+    if (performanceEnabled && !isRunningTransaction(activity)) {
+      final ITransaction transaction =
+          hub.startTransaction(getActivityName(activity), "navigation");
+
+      // lets bind to the scope so other integrations can pick it up
+      hub.configureScope(
+          scope -> {
+            // we we'd not like to overwrite existent transactions bound to the Scope manually.
+            if (scope.getTransaction() == null) {
```

Comment:
`h`: I think we will run into trouble with putting transactions on the scope. I think we can't use that to give the users the possibility to finish the transaction manually in the activity. We need another way to store and expose the transaction for the current activity.

Reply:
> `h`: I think we will run into trouble with putting transactions on the scope. I think we can't use that to give the users the possibility to finish the transaction manually in the activity. We need another way to store and expose the transaction for the current activity.

what type of troubles? can you elaborate?

Reply:
If we don't put the transaction on the scope, user code or our own integrations can't find it and no spans will be added to it.

Reading and writing isn't an atomic operation here and scope is accessed concurrently.
One alternative is a single method that can do the condition and reference swap atomically like `scope.setIfNull(transaction)` but the fact we got to this situation is a sign we probably need to discuss this further. In other SDKs/general Sentry unified API a scope is part of a hub which is usually accessed by a single thread (thread local storage). It must be thread-safe as it can be accessed concurrently particularly when it's being cloned to make room for a new instance bound to a new thread, but it's generally not being mutated by multiple threads at the same time. This isn't true for the mobile SDKs since we wanted to have this global state for breadcrumbs and tags etc.

For transactions this seems not to fit very well so we might need to discuss what to do next here.

Reply:
that's a good point, we do it atomically for `startSession` for example, maybe this would be a workaround for making it work now.